src/aspire/reconstruction/mean.py

```diff
@@ -1,17 +1,262 @@
+        Weighted mean volume estimation.
+
+        This is best considered as an r x r matrix of volumes;
+        each volume is a weighted mean least-squares estimator kernel (MeanEstimator).
+        Convolution with each of these kernels is equivalent
+        to performing a projection/backprojection on a volume,
+        with the appropriate amplitude modifiers and CTF,
+        and also a weighting term;
+        the r^2 volumes are each of pairwise products between the weighting vectors given by the columns of wts.
+
+        Note that this is a non-centered Fourier transform, so the zero frequency is found at index 0.
+
+        Formulas and conventions used for Volume estimation are described in:
+
+        “Structural Variability from Noisy Tomographic Projections.”
+        Andén, Joakim, and Amit Singer.
+        SIAM journal on imaging sciences vol. 11,2 (2018): 1441-1492.
+        doi:10.1137/17M1153509
+
+        "Cryo-EM reconstruction of continuous heterogeneity by Laplacian spectral volumes"
+        Amit Moscovich, Amit Halevi, Joakim Andén and Amit Singer
+        Inverse Problems, Volume 36, Number 2, 2020 IOP Publishing Ltd
+        Special Issue on Cryo-Electron Microscopy and Inverse Problems
+        https://doi.org/10.1088/1361-6420/ab4f55
+
+        :param weights: Matrix of weights, n x r.
+        """
+
+        self.weights = weights
+        self.r = self.weights.shape[1]
+        super().__init__(*args, **kwargs)
+        assert self.src.n == self.weights.shape[0]
+
+    def __getattr__(self, name):
+        if name == "precond_kernel":
+            if self.preconditioner == "circulant":
+                # TODO: Discuss precond plans.
+                # self.precond_kernel = FourierKernelMat(
+                #     1.0 / self.kernel.circularize(), centered=True
+                # )
+                raise NotImplementedError(
+                    "Circulant preconditioner not implemented for WeightedVolumesEstimator."
+                )
+            else:
+                if self.preconditioner.lower() not in (None, "none"):
+                    logger.warning(
+                        f"Preconditioner {self.preconditioner} is not implemented, resetting to default of None."
+                    )
+                self.precond_kernel = None
+            return self.precond_kernel
+
+        else:
+            #  raise AttributeError(name)
+            return super().__getattr__(name)
+
+    def compute_kernel(self):
+        """
+        :return: r x r matrix of volumes, shaped (r, r, 2L, 2L, 2L).
+        """
+
+        _2L = 2 * self.src.L
+        # Note, because we're iteratively summing it is critical we zero this array.
+        kernel = np.zeros((self.r, self.r, _2L, _2L, _2L), dtype=self.dtype)
+        sq_filters_f = np.square(evaluate_src_filters_on_grid(self.src))
+
+        for k in range(self.r):
+            for j in range(k + 1):
+                for i in range(0, self.src.n, self.batch_size):
```

Comment:
Wondering if this should be the outer loop since the first half-dozen lines do not depend on `k` or `j`.

Reply:
Makes sense, sorry I missed that before.

Reply:
Great. But we're still recalculating grids, filters, amplitudes, etc. for each `j, k`, which is unnecessary. Everything up to
```
                    weights *= (
                        self.weights[_range, j] * self.weights[_range, k]
                    ).reshape(1, 1, len(_range))
```
can be moved outside the `j, k` loops.

Reply:
Right.  Very sorry.  Don't know how I missed that in your first suggestion. Must have been distracted.

I went ahead and factored out the calcs you mentioned.

Felt a little guilty, so I've also refactored `compute_kernel` into a shared `_compute_kernel`.  It looks like the only difference besides the dimension was a scaling term at the `anufft` line.  Using `r` in the scaling term it seems to generalize.